### PR TITLE
Remove INCLUDE_LD_SO_LIST_DIAGNOSTICS

### PR DESCRIPTION
**Pull request by jankratochvil**
It was introduced in glibc-2.34 together with preferred CPU_FEATURE_ACTIVE.

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Reviewers
 * [Radim Vansa](https://openjdk.org/census#rvansa) (@rvansa - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/crac.git pull/136/head:pull/136` \
`$ git checkout pull/136`

Update a local copy of the PR: \
`$ git checkout pull/136` \
`$ git pull https://git.openjdk.org/crac.git pull/136/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 136`

View PR using the GUI difftool: \
`$ git pr show -t 136`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/crac/pull/136.diff">https://git.openjdk.org/crac/pull/136.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/crac/pull/136#issuecomment-1786565042)

### CODE REVIEW

File: src/hotspot/cpu/x86/vm_version_x86.cpp [hunk 1]
```diff
@@ -855,8 +853,7 @@ uint64_t VM_Version::CPUFeatures_parse(uint64_t &glibc_features) {
   return -1;
 }
 
-#if INCLUDE_LD_SO_LIST_DIAGNOSTICS
-
+//#if INCLUDE_LD_SO_LIST_DIAGNOSTICS
```
Comment (rvansa) [inline src/hotspot/cpu/x86/vm_version_x86.cpp:856]: Please remove commented out code.

File: src/hotspot/cpu/x86/vm_version_x86.cpp [hunk 2]
```diff
@@ -942,12 +941,12 @@ bool VM_Version::glibc_env_set(char *disable_str) {
   char *env_val = disable_str;
   const char *env = getenv(TUNABLES_NAME);
   if (env && strcmp(env, env_val) == 0) {
-#if !INCLUDE_CPU_FEATURE_ACTIVE && !INCLUDE_LD_SO_LIST_DIAGNOSTICS
-    if (ShowCPUFeatures) {
-      tty->print_cr("Environment variable already set, both glibc CPU_FEATURE_ACTIVE and ld.so --list-diagnostics are unavailable - re-exec suppressed: " TUNABLES_NAME "=%s", env);
+    if (!INCLUDE_CPU_FEATURE_ACTIVE && !INCLUDE_LD_SO_LIST_DIAGNOSTICS) {
```
Comment (rvansa) [inline src/hotspot/cpu/x86/vm_version_x86.cpp:944]: Have you tested compilation with these disabled? Looks to me like the compiler could complain about unreachable code when the condition is always false...

**Review by rvansa — COMMENTED** [review]
While I like the certainty that the code is always compilable, I wonder if handling optional features like this (using macros in conditions) is a pattern that is used elsewhere in the JDK?

File: src/hotspot/cpu/x86/vm_version_x86.cpp (hunk 1 above)
Comment (jankratochvil) [reply src/hotspot/cpu/x86/vm_version_x86.cpp:856]: I wanted to indicate this function is used only for `INCLUDE_LD_SO_LIST_DIAGNOSTICS`. I can also put there a full sentence.

File: src/hotspot/cpu/x86/vm_version_x86.cpp (hunk 2 above)
Comment (jankratochvil) [reply src/hotspot/cpu/x86/vm_version_x86.cpp:944]: I have tried all the 3 compilation cases only now. I do not see such compiler warnings (although different compilations and compilers produce different warnings) but I could fix some other problems. Which will make me update #112.

**Comment by jankratochvil** [comment]
> While I like the certainty that the code is always compilable, I wonder if handling optional features like this (using macros in conditions) is a pattern that is used elsewhere in the JDK?

I haven't found such case. I expect to leave such decisions up to OpenJDK reviewers during CRaC merge request.

OTOH to make things easier - the whole `INCLUDE_LD_SO_LIST_DIAGNOSTICS` code should be probably removed. The support for `INCLUDE_LD_SO_LIST_DIAGNOSTICS` will probably never be used as it was added to glibc only few months before the (better) `INCLUDE_CPU_FEATURE_ACTIVE` support. Before doing such removal for real one should check whether any major/supported Linux distribution has been released with such glibc version supporting only `INCLUDE_LD_SO_LIST_DIAGNOSTICS` without `INCLUDE_CPU_FEATURE_ACTIVE`. I unfortunately did not check what was when implemented in glibc while writing the `INCLUDE_LD_SO_LIST_DIAGNOSTICS` code.

**Comment by rvansa** [comment]
It's in our interest to provoke as few discussions as possible, to get CRaC merged without extra delay. So, the code style is questionable and there's actually very little chance it will be useful, do you still consider this worth merging? Or would you prefer to remove it altogether instead? (Could be done in this very PR)

**Comment by jankratochvil** [comment]
So if there is the decision to remove it I will investigate first there is really no interesting Linux OS release with the "problematic" glibc version.

File: src/hotspot/cpu/x86/vm_version_x86.cpp [hunk 3]
```diff
@@ -944,9 +862,9 @@ bool VM_Version::glibc_env_set(char *disable_str) {
   char *env_val = disable_str;
   const char *env = getenv(TUNABLES_NAME);
   if (env && strcmp(env, env_val) == 0) {
-    if (!INCLUDE_CPU_FEATURE_ACTIVE && !INCLUDE_LD_SO_LIST_DIAGNOSTICS) {
+    if (!INCLUDE_CPU_FEATURE_ACTIVE) {
       if (ShowCPUFeatures) {
-        tty->print_cr("Environment variable already set, both glibc CPU_FEATURE_ACTIVE and ld.so --list-diagnostics are unavailable - re-exec suppressed: " TUNABLES_NAME "=%s", env);
+        tty->print_cr("Environment variable already set, both glibc CPU_FEATURE_ACTIVE is unavailable - re-exec suppressed: " TUNABLES_NAME "=%s", env);
```
Comment (rvansa) [inline src/hotspot/cpu/x86/vm_version_x86.cpp:867]: The "both" word is extra in here I guess.

**Comment by rvansa** [comment]
I admit I have no idea about the IBT and SHSTK bits; it might be beneficial to include the description of what you're changing about those in the PR description so that whoever stumbles upon the commit has some pointers.

**Comment by jankratochvil** [comment]
> I admit I have no idea about the IBT and SHSTK bits; it might be beneficial to include the description of what you're changing about those in the PR description so that whoever stumbles upon the commit has some pointers.

That is an unrelated change. So I will split it to another pull request.

**Review by rvansa — APPROVED** [review]
LGTM, thanks!

**Pull request closed without merging**